src/mapreduce.jl

```diff
@@ -29,7 +16,7 @@ end
     newT = :(Core.Inference.return_type(f, Tuple{$(eltypes...)}))
     return quote
         @_inline_meta
-        @inbounds return similar_type(typeof(_first(a...)), $newT)(tuple($(exprs...)))
+        @inbounds return similar_type(typeof(_first_static(a...)), $newT, Size(S))(tuple($(exprs...)))
```

Comment:
Is `_first_static` the ideal choice here? (Just asking - I'm not sure what the answer is).

Until/unless we have a proper promotion mechanism, having some memorable rule like `similar` and `similar_type` being called on the first argument has certain advantages. 

It's hard to predict what the user wants here, and we don't typically have the option of `map!` for when the user wants to make the choice explicitly, so perhaps a certain API for `map` (or `map_to_type`, etc) could be useful.